### PR TITLE
feat: expose metrics for user accounts per member cluster

### PR DESCRIPTION
**Pull request by xcoulon**
also, introduce a new 'indexOfMember' func

e2e: https://github.com/codeready-toolchain/toolchain-e2e/pull/245

Fixes https://issues.redhat.com/browse/CRT-970
Signed-off-by: Xavier Coulon <xcoulon@redhat.com>

### CODE REVIEW

File: pkg/controller/toolchainstatus/toolchainstatus_controller.go [hunk 1]
```diff
@@ -306,6 +306,7 @@ func (r *ReconcileToolchainStatus) registrationServiceHandleStatus(reqLogger log
 // if any of the members are not ready or if no member clusters are found
 func (r *ReconcileToolchainStatus) membersHandleStatus(logger logr.Logger, toolchainStatus *toolchainv1alpha1.ToolchainStatus) bool {
 	// get member clusters
+	log.Info("updating member status")
```
Comment (alexeykazakov) [inline pkg/controller/toolchainstatus/toolchainstatus_controller.go:309]:
```suggestion
	logger.Info("updating member status")
```

**Review by alexeykazakov — APPROVED** [review]
Looks good! Just a minor comment to address.

File: pkg/metrics/metrics.go [hunk 2]
```diff
@@ -54,6 +62,8 @@ func initMetrics() {
 	UserSignupAutoDeactivatedTotal = newCounter("user_signups_auto_deactivated_total", "Total number of Automatically Deactivated User Signups")
 	// Gauges
 	MasterUserRecordGauge = newGauge("master_user_record_current", "Current number of Master User Records")
+	// GaugeVecs
+	MasterUserRecordGaugeVec = newGaugeVec("master_user_records_current", "Current number of Master User Records", "cluster_name")
```
Comment (MatousJobanek) [inline pkg/metrics/metrics.go:66]: it's very minor and you can ignore it, but it would be fine if the var name was similar to the name of the metric:
```suggestion
	MasterUserRecordsGaugeVec = newGaugeVec("master_user_records_current", "Current number of Master User Records", "cluster_name")
```

File: pkg/metrics/metrics_test.go [hunk 3]
```diff
@@ -45,17 +45,22 @@ func TestRegisterCustomMetrics(t *testing.T) {
...
 	UserSignupUniqueTotal.Inc()
 	MasterUserRecordGauge.Set(22)
+	MasterUserRecordGaugeVec.WithLabelValues("member-1").Set(20)
 
 	Reset()
 
 	// then
 	assert.Equal(t, float64(0), promtestutil.ToFloat64(UserSignupUniqueTotal))
 	assert.Equal(t, float64(0), promtestutil.ToFloat64(MasterUserRecordGauge))
+	assert.Equal(t, float64(0), promtestutil.ToFloat64(MasterUserRecordGaugeVec.WithLabelValues("member-1")))
```
Comment (MatousJobanek) [inline pkg/metrics/metrics_test.go:65]: since every metric type has also its own unit test verifying the number that is being set: https://github.com/codeready-toolchain/host-operator/blob/53c13926ba73b3d631a9ac2d21bc4ec6ba8d13ac/pkg/metrics/metrics_test.go#L12-L32
could you please add one also for GaugeVec?

File: test/counter.go [hunk 4]
```diff
@@ -27,23 +27,23 @@ func UserAccountsForCluster(clusterName string, number int) ExpectedNumberOfUser
 func AssertThatUninitializedCounterHas(t *testing.T, numberOfMurs int, numberOfUasPerCluster ...ExpectedNumberOfUserAccounts) {
 	counts, err := counter.GetCounts()
 	assert.EqualErrorf(t, err, "counter is not initialized", "should be error because counter hasn't been initialized yet")
-
 	verifyCounts(t, counts, numberOfMurs, numberOfUasPerCluster...)
 }
 
 func AssertThatCounterHas(t *testing.T, numberOfMurs int, numberOfUasPerCluster ...ExpectedNumberOfUserAccounts) {
-	AssertMetricsGaugeEquals(t, numberOfMurs, metrics.MasterUserRecordGauge)
```
Comment (MatousJobanek) [inline test/counter.go:35]: May I know why did you remove it?

**Review by MatousJobanek — COMMENTED** [review]
Nice & simple :+1:

File: pkg/metrics/metrics.go (hunk 2 above)
Comment (xcoulon) [reply pkg/metrics/metrics.go]: actually, I renamed it to `UserAccountGaugeVec` since it is associated with the current number of UserAccount resources

File: pkg/metrics/metrics_test.go (hunk 3 above)
Comment (xcoulon) [reply pkg/metrics/metrics_test.go]: done in 1475e93

File: test/counter.go (hunk 4 above)
Comment (xcoulon) [reply test/counter.go:35]: I moved it in the `verifyCounts` func which is called on line 36, so the assertion on the `metrics.MasterUserRecordGauge` is done when calling both the `AssertThatCounterHas` and the `AssertThatUninitializedCounterHas` funcs

**Review by MatousJobanek — APPROVED** [review]
